astronomer/providers/sftp/triggers/sftp.py

```diff
@@ -0,0 +1,63 @@
+import asyncio
+from typing import Any, AsyncIterator, Dict, Tuple
+
+from airflow.exceptions import AirflowException
+from airflow.triggers.base import BaseTrigger, TriggerEvent
+
+from astronomer.providers.sftp.hooks.sftp import SFTPHookAsync
+
+
+class SFTPTrigger(BaseTrigger):
+    """
+    Trigger that fires when either the path on the SFTP server does not exist,
+    or when there are no files matching the file pattern at the path
+
+    :param path: The path on the SFTP server to search for a file matching the file pattern.
+                Authentication method used in the SFTP connection must have access to this path
+    :param file_pattern: Pattern to be used for matching against the list of files at the path above.
+                Uses the fnmatch module from std library to perform the matching.
+
+    :param sftp_conn_id: SFTP connection ID to be used for connecting to SFTP server
+    :param poke_interval: How often, in seconds, to check for the existence of the file on the SFTP server
+    """
+
+    def __init__(
+        self,
+        path: str,
+        file_pattern: str = "",
+        sftp_conn_id: str = "sftp_default",
+        poke_interval: float = 5,
+    ) -> None:
+        super().__init__()
+        self.path = path
+        self.file_pattern = file_pattern
+        self.sftp_conn_id = sftp_conn_id
+        self.poke_interval = poke_interval
+
+    def serialize(self) -> Tuple[str, Dict[str, Any]]:
+        """Serializes SFTPTrigger arguments and classpath"""
+        return (
+            "astronomer.providers.sftp.triggers.sftp.SFTPTrigger",
+            {
+                "path": self.path,
+                "file_pattern": self.file_pattern,
+                "sftp_conn_id": self.sftp_conn_id,
+                "poke_interval": self.poke_interval,
+            },
+        )
+
+    async def run(self) -> AsyncIterator["TriggerEvent"]:  # type: ignore[override]
+        """
+        Makes a series of asynchronous calls to sftp servers via async sftp hook. It yields a Trigger if
+        file matching file pattern exists at the specified path, otherwise it throws an exception.
+        """
+        hook = self._get_async_hook()
+        while True:
+            try:
+                await hook.get_file_by_pattern(path=self.path, fnmatch_pattern=self.file_pattern)
+                yield TriggerEvent(True)
+            except AirflowException:
+                await asyncio.sleep(self.poke_interval)
```

Comment:
Also, I think if a file is not found then only we should sleep in case of any other error we should throw error. WDYT?

Reply:
This is how it works now, albeit in a roundabout way.

In the hook
```
        try:
            files = await sftp_client.listdir(path)
            return sorted(files)
        except asyncssh.SFTPNoSuchFile:
            raise AirflowException(f"No files at path {path} found — Deferring")
```

we only catch the exception when a file isn't found. This means that any connection errors will fall through and halt the sensor, throwing the raw exception back to Airflow (which isn't then caught, because it's not an `AirflowException`)

Reply:
Ok, In that case, any other exception than `AirflowException` we should yield the error something like
```
yield TriggerEvent({"status": "error", message: e})
```
then if `status` is an error then raise an exception in SFTPSensorAsync  `execute_complete` method you can access it like
```
if event["status"] == "error":
    raise
```

Reply:
Got it, this makes more sense. Going to rework this a bit so that `AirflowExceptions` aren't being used as heavily for control flow. It was causing the triggerer to get stuck in a loop after the task had finished/failed execution.